ride-service/src/main/kotlin/com/dulfinne/taxi/rideservice/service/impl/PassengerServiceImpl.kt

```diff
@@ -0,0 +1,142 @@
+    }
+
+    @Transactional
+    override fun createRide(passengerId: Long, request: LocationRequest): RideResponse {
+        val ride = mapper.toRide(request, passengerId)
+        ride.price = countPriceByLocation(ride.startPosition, ride.endPosition)
+
+        repository.save(ride)
+        return mapper.toRideResponse(ride)
+    }
+
+    @Transactional
+    override fun cancelRide(rideId: Long, passengerId: Long) {
+        val ride = getRideIfExists(rideId)
+        validatePassenger(ride, passengerId)
+
+        if (ride.status != RideStatus.SEARCHING.id) {
+            throw ActionNotAllowedException(ExceptionKeys.CANCEL_NOT_ALLOWED, RideStatus.fromId(ride.status))
+        }
+
+        ride.status = RideStatus.CANCELLED.id
+        repository.save(ride)
+    }
+
+    @Transactional(readOnly = true)
+    override fun rateDriver(rideId: Long, passengerId: Long, request: RatingRequest) {
+        val ride = getRideIfExists(rideId)
+        validatePassenger(ride, passengerId)
+
+        if (ride.status != RideStatus.COMPLETED.id) {
+            throw ActionNotAllowedException(ExceptionKeys.RATE_NOT_ALLOWED, RideStatus.fromId(ride.status))
+        }
+        checkRideCanBeRated(ride.endTime!!)
+
+        //TODO: send ratingModel to driver-service
+    }
+
+    @Transactional(readOnly = true)
+    override fun getAllPassengerRides(
+        passengerId: Long,
+        offset: Int,
+        limit: Int,
+        sortField: String
+    ): Page<RideResponse> {
+
+        val ridesPage = repository.findAllByPassengerId(
+            passengerId,
+            PageRequest.of(offset, limit, Sort.by(Sort.Direction.ASC, sortField))
+        )
+        return ridesPage.map(mapper::toRideResponse)
+    }
+
+    @Transactional(readOnly = true)
+    override fun getRideById(passengerId: Long, rideId: Long): RideResponse {
+        val ride = getRideIfExists(rideId)
+        validatePassenger(ride, passengerId)
+        return mapper.toRideResponse(ride)
+    }
+
+    private fun countPriceByLocation(startPosition: Point, endPosition: Point): BigDecimal {
+        val basePrice = RideConstants.BASE_PRICE
+        val pricePerKm = RideConstants.PRICE_PER_KM
+
+        val distanceKm = haversineDistance(startPosition.x, startPosition.y, endPosition.x, endPosition.y)
+        val totalPrice = basePrice.add(distanceKm.multiply(pricePerKm))
+
+        return totalPrice.setScale(2, RoundingMode.HALF_UP)
+    }
+
+    fun haversineDistance(lat1: Double, lon1: Double, lat2: Double, lon2: Double): BigDecimal {
+        val earthRadiusKm = 6371.0
+        val diffLat = Math.toRadians(lat2 - lat1)
+        val diffLon = Math.toRadians(lon2 - lon1)
+
+        val a = sin(diffLat / 2).pow(2.0) + cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(diffLon / 2).pow(2.0)
+        val c = 2 * atan2(sqrt(a), sqrt(1 - a))
+
+        val distanceKm = earthRadiusKm * c
+        return BigDecimal(distanceKm).setScale(6, RoundingMode.HALF_UP)
+    }
```

Comment:
Wow! Good for you. Did the same thing, when I was in EPAM, calculating the distance between two points taking into consideration the earth radius.

Respect and like for that